### DISCUSSION TITLE
Any plans for PK chunking support?

### DISCUSSION THREAD

**Discussion by bdmayes** [jsforce/jsforce#1433] [Ideas]
Are there any plans to support PK chunking via the Bulk API? This technique is mainly used to query object types with many millions of records and is described here:

https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/async_api_headers_enable_pk_chunking.htm

It's obviously only available for certain standard objects, but it appears to support all custom objects as well. By sending the Sforce-Enable-PKChunking header with the API request, Salesforce will automatically use the primary key of the table (Id) to split the batch into smaller batches and query a specific range of base-62 Ids based on chunk size. Since this is a primary key and therefore extremely well-indexed, it allows queries on large data sets to proceed extremely quickly.

I see no option to pass this into the Bulk implementation in jsforce (neither high or low level). It seems like this could be implemented in the low-level implementation by calling `Bulk#createJob()` with `operation` set to "query" and then creating a batch from that job with the appropriate header(s) sent across in the POST (note that you can pass over an optional startRow if you're retrying/resuming a previously failed job).

It appears that creating a batch with the PK chunking header will force Salesforce to create many sub-batches, each of which can be monitored and retrieved separately. Once all batches have been downloaded, the job can be closed.

I don't mind implementing this using the low-level API on my own, but I don't even see any support for passing in those headers into a bulk job. Do you have any plans to support this in the near future? Thanks!

**Comment by ryansieve** [comment]
+1

**Comment by david0673** [comment]
I've proposed a PR to add support for this feature, please take a look & comment if you feel there is something missing or to be improved

https://github.com/jsforce/jsforce/pull/780

**Comment by cristiand391** [comment]
The Bulk 2.0 API enables PKChunking by default unless you use ` ORDER BY`or `LIMIT`:
> Don’t use ORDER BY or LIMIT, as they disable PKChunking for the query. With PKChunking disabled, the query takes longer to execute, and potentially results in query timeouts. If ORDER BY or LIMIT is used, and you experience query time outs, then remove the ORDER BY or LIMIT clause before any subsequent troubleshooting.

https://developer.salesforce.com/docs/atlas.en-us.api_asynch.meta/api_asynch/query_create_job.htm

If you are on v3 you should be able to migrate from `conn.bulk` to `conn.bulk2`:
https://jsforce.github.io/document/#bulk-v2-api
